addons/account/models/account_move.py

```diff
@@ -4703,277 +4703,336 @@ def _mail_track(self, tracked_fields, initial):
-                            remaining_credit_amount -= credit_exchange_amount
-
-                if exchange_lines_to_fix:
-                    exchange_vals = exchange_lines_to_fix._prepare_exchange_difference_move_vals(
-                        amounts_list,
-                        exchange_date=max(debit_line.date, credit_line.date),
-                    )
-                    exchange_data[len(partials_vals_list)] = exchange_vals
+        remaining_debit_amount -= partial_amount
+        remaining_credit_amount += partial_amount
+        remaining_debit_amount_curr -= partial_debit_amount_currency
+        remaining_credit_amount_curr += partial_credit_amount_currency
+
+        res['partial_vals'] = {
+            'amount': partial_amount,
+            'debit_amount_currency': partial_debit_amount_currency,
+            'credit_amount_currency': partial_credit_amount_currency,
+            'debit_move_id': debit_vals.get('record') and debit_vals['record'].id,
+            'credit_move_id': credit_vals.get('record') and credit_vals['record'].id,
+        }
 
-            # ==== Create partials ====
+        debit_vals['amount_residual'] = remaining_debit_amount
+        debit_vals['amount_residual_currency'] = remaining_debit_amount_curr
+        credit_vals['amount_residual'] = remaining_credit_amount
+        credit_vals['amount_residual_currency'] = remaining_credit_amount_curr
 
-            remaining_debit_amount -= partial_amount
-            remaining_credit_amount += partial_amount
-            remaining_debit_amount_curr -= partial_debit_amount_currency
-            remaining_credit_amount_curr += partial_credit_amount_currency
+        if debit_fully_matched:
+            res['debit_vals'] = None
+        if credit_fully_matched:
+            res['credit_vals'] = None
+        return res
 
-            partials_vals_list.append({
-                'amount': partial_amount,
-                'debit_amount_currency': partial_debit_amount_currency,
-                'credit_amount_currency': partial_credit_amount_currency,
-                'debit_move_id': debit_line.id,
-                'credit_move_id': credit_line.id,
-            })
+    @api.model
+    def _prepare_reconciliation_partials(self, vals_list):
+        ''' Prepare the partials on the current journal items to perform the reconciliation.
+        Note: The order of records in self is important because the journal items will be reconciled using this order.
+
+        :return: a tuple of 1) list of vals for partial reconciliation creation, 2) the list of vals for the exchange difference entries to be created
+        '''
+        debit_vals_list = iter([x for x in vals_list if x['balance'] > 0.0 or x['amount_currency'] > 0.0])
+        credit_vals_list = iter([x for x in vals_list if x['balance'] < 0.0 or x['amount_currency'] < 0.0])
+        debit_vals = None
+        credit_vals = None
+
+        partials_vals_list = []
+        exchange_data = {}
+
+        while True:
```

Comment:
It could be 
```py
while debit_vals and credit_vals
```
You just need to initialize it with a first `whatever_vals = next(whatever_vals_list, None)` and proceed to re-assignment at the end of the loop instead of assigning `None`